src/sources/statsd/mod.rs

```diff
@@ -487,8 +495,36 @@ mod test {
             .collect::<AbsoluteMetricState>();
         let metrics = state.finish();
 
-        assert_counter(&metrics, series!("foo", "a" => "true", "b" => "b"), 100.0);
-        assert_counter(&metrics, series!("foo", "a" => "true", "b" => "c"), 100.0);
+        let series_one = MetricSeries {
+            name: MetricName {
+                name: "foo".to_owned(),
+                namespace: None,
+            },
+            tags: Some(
+                BTreeMap::from([
+                    ("a".to_owned(), TagValue::from(None)),
+                    ("b".to_owned(), TagValue::from(Some("b".to_owned()))),
+                ])
+                .into(),
```

Comment:
It doesn't compile unfortunately. I did try to use the existing macros during the implementation, though I might have missed something with that, hence my note in the opening comment / slack.

I am a macro newb but would love to be on the review for the PR that tackles that, as I have the context to why that change is needed.

```
the trait bound `std::string::String: std::convert::From<TagValueSet>` is not satisfied
```

Reply:
Ah, yeah. So the macro:
https://github.com/vectordotdev/vector/blob/master/lib/vector-core/src/event/metric/mod.rs#L45-L47

essentially expands to:

```rust
MetricTags::from([
  ("a".into(), TagValueSet::Empty.into())
])
```

But there is only an [impl](https://github.com/vectordotdev/vector/blob/master/lib/vector-core/src/event/metric/tags.rs#L530) for 

```rust
impl<const N: usize> From<[(String, String); N]> for MetricTags {
```

So all the keys and values have to have an `into` to `String`. Hence the compile error.

You could change the macro to be something like:

```rust
    ($($key:expr => $value:expr),*) => {
        [
            $( ($key.into(), $crate::event::metric::TagValue::from($value)), )*
        ].into_iter().collect::<$crate::event::MetricTags>()
    };
```

So it will then fall back to: 

```rust
impl FromIterator<(String, TagValue)> for MetricTags {
```

Then the value can be anything that has an Into for `TagValue`.

You then need to add a couple of extra `Impl`s for this to cover some extra cases:

```rust
impl From<&str> for TagValue {
    fn from(value: &str) -> Self {
        Self::Value(value.to_string())
    }
}

impl From<Cow<'_, str>> for TagValue {
    fn from(value: Cow<'_, str>) -> Self {
        Self::Value(value.to_string())
    }
}
```

And then it all seems to compile.

Reply:
Thanks for taking the time to provide those details! I will take a look at adding this.

Reply:
Works like a charm!